[WIP] JSON Coding Keys - string keyed dictionary opt out

mortenbekditlevsen opened the pull request:
<!-- What's in this pull request? -->
Using key conversion strategies for JSONEncoder and JSONDecoder also converted keys of Dictionary values in the Codable models. This PR implements custom boxing and unboxing for String keyed Dictionaries.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves a bug with key coding of dictionary keys discussed here:
https://forums.swift.org/t/jsondecoder-keydecodingstrategy-issue-with-dictionary/11788/9

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

mortenbekditlevsen committed:
Make JSONEncoder and JSONDecoder circumvent keyEncodingStrategies and keyDecodingStrategies for String keyed Dictionaries.

mortenbekditlevsen committed:
Merge branch 'master' of https://github.com/apple/swift

mortenbekditlevsen committed:
Remove unused function

itaiferber commented on stdlib/public/SDK/Foundation/JSONEncoder.swift:
```diff
@@ -815,30 +823,45 @@ extension _JSONEncoder {
         return try self.box_(value) ?? NSDictionary()
     }
 
+    fileprivate func box(_ dict: [String : Encodable]) throws {
```
I think this would be a bit cleaner if it were handled like the other `box(_:)` overloads — this can return an `NSObject` (`NSDictionary` under the hood), and the type can be handled in the `if` like any other case

mortenbekditlevsen replied:
In the latest change, I've tried making it look more like the existing `box(_:)` overloads.

itaiferber commented on stdlib/public/SDK/Foundation/JSONEncoder.swift:
```diff
@@ -815,30 +823,45 @@ extension _JSONEncoder {
+        } else if type == Decimal.self || type == NSDecimalNumber.self {
             // JSONSerialization can natively handle NSDecimalNumber.
             return (value as! NSDecimalNumber)
         }
```
We can probably move up the box call from below to up here. Performance-wise, I don't know how expensive it would be to check `if let stringKeyedDict = value as? [String : Encodable]` vs. adding another marker protocol and doing the protocol check, but this would be something good to test.

mortenbekditlevsen replied:
I've added another marker protocol. I don't know how to go about testing the performance of the cast vs. a protocol check, but now the marker is there and the cast is only performed if the marker is present.

itaiferber commented on stdlib/public/SDK/Foundation/JSONEncoder.swift:
```diff
@@ -2359,11 +2382,29 @@ extension _JSONDecoder {
 
+    fileprivate func unbox<T>(_ value: Any, as type: _JSONStringDictionaryMarker.Type) throws -> T? {
+        var result: Dictionary<String, Any> = [:]
+        let dict = value as! Dictionary<String, NSObject>
```
We can't `as!` here as we need to confirm that the value we have does actually match the request. We need to get

```swift
guard let dict = value as? NSDictionary else {
    // throw error; see other type mismatch examples throughout
}
```

or else we'd crash whenever you request a dictionary but something else is present instead.

We can also iterate through the `NSDictionary` directly instead of casting to `[String : NSObject]`; we shouldn't pay the cost of an `O(n)` bridge.

mortenbekditlevsen replied:
Ah, right - I mixed up the value vs. request concepts here.
Also did forgot to consider that the types are already NSDictionaries from the JSONSerialization.

When considering the NSDictionary I also have to ensure that the keys are Strings. Is it ok here to throw a similar type mismatch error on the keys not being Strings?

itaiferber replied:
You don't need to worry about this case — `JSONSerialization` can only ever return a dictionary with string keys.

itaiferber commented on stdlib/public/SDK/Foundation/JSONEncoder.swift:
```diff
@@ -2359,11 +2382,29 @@ extension _JSONDecoder {
     }
 
+    fileprivate func unbox<T>(_ value: Any, as type: _JSONStringDictionaryMarker.Type) throws -> T? {
+        var result: Dictionary<String, Any> = [:]
```
Style-wise, I think we'd prefer `var result = [String : Any]()`

itaiferber reviewed (changes requested):
Thanks so much for putting this together! I think this looks good overall; we just need to make a few changes.

I would be curious to see too if there's a clear difference performance-wise between checking whether a value is a string-keyed dictionary on encode using another marker protocol vs. just trying to cast the value.

itaiferber commented:
There will also now be some merge conflicts, so if you don't mind rebasing, we'll be able to see how these changes apply.

mortenbekditlevsen committed:
Merge branch 'master' of https://github.com/apple/swift

mortenbekditlevsen committed:
Merge branch 'master' into jsoncodingkeys

mortenbekditlevsen committed:
Implemented fixes suggested by Itai

mortenbekditlevsen committed:
Ensure that keys are strings

mortenbekditlevsen committed:
Added another marker protocol for encodable string keyed dictionaries

mortenbekditlevsen commented:
Hi Itai,
I have merged the latest master into my branch, but I'm afraid that I have messed up something because running update-checkout now fails for me during a rebase. This also happens when I check out the master branch of the swift upstream repo, so I can't really figure out what I am doing wrong.

So for these latest commits I haven't been able to either compile or test my changes at all. Many apologies if things don't work at all.

I've implemented your suggestions as best as I could - but I haven't been able to do a performance test of the marker vs. cast.

Should I start over with a fresh checkout, cherry-pick the changes and do a new PR?

mortenbekditlevsen committed:
Merge branch 'master' into jsoncodingkeys

mortenbekditlevsen committed:
Imitate container stack logic from `box_(_ value: Encodable)` in `box(_ dict: [String : Encodable])`

mortenbekditlevsen commented:
And now everything builds and tests succeed again.

itaiferber commented on stdlib/public/SDK/Foundation/JSONEncoder.swift:
```diff
@@ -10,6 +10,19 @@
 //===----------------------------------------------------------------------===//
 
+
+fileprivate protocol _JSONStringDictionaryEncodableMarker { }
```
Some doc comments about the presence of these protocols would be helpful. 🙂 
Something along the lines of `"A marker protocol used to determine whether a value is a String-keyed dictionary (in which case it should be exempt from key conversion strategies)."` would be great. (Similarly for the decodable marker as well.)

itaiferber commented on stdlib/public/SDK/Foundation/JSONEncoder.swift:
```diff
@@ -815,24 +828,53 @@ extension _JSONEncoder {
+        let depth = self.storage.count
+        let result = self.storage.pushKeyedContainer()
+        do {
+            for (k, v) in dict {
```
Slight style nit: we prefer `(key, value)` to `(k, v)` where possible (when names don't conflict)

itaiferber commented on stdlib/public/SDK/Foundation/JSONEncoder.swift:
```diff
@@ -815,24 +828,53 @@ extension _JSONEncoder {
             // JSONSerialization can natively handle NSDecimalNumber.
             return (value as! NSDecimalNumber)
+        } else if value is _JSONStringDictionaryEncodableMarker {
+            return try self.box(value as! Dictionary<String, Encodable>)
```
Another slight style nit: we prefer `[String : Encodable]`

itaiferber commented on stdlib/public/SDK/Foundation/JSONEncoder.swift:
```diff
@@ -2359,11 +2401,34 @@ extension _JSONDecoder {
+        }
+        let elementType = type.elementType
+        for (k, v) in dict {
+            guard let key = k as? String else {
```
You can `as!` here; `[NS]JSONSerialization` won't return a dictionary with anything but string keys

itaiferber commented on stdlib/public/SDK/Foundation/JSONEncoder.swift:
```diff
@@ -2359,11 +2401,34 @@ extension _JSONDecoder {
         }
     }
 
+    fileprivate func unbox<T>(_ value: Any, as type: _JSONStringDictionaryDecodableMarker.Type) throws -> T? {
```
Please check for `NSNull` here like in the other methods above

itaiferber commented on stdlib/public/SDK/Foundation/JSONEncoder.swift:
```diff
@@ -815,24 +828,53 @@ extension _JSONEncoder {
     }
 
-    fileprivate func box<T : Encodable>(_ value: T) throws -> NSObject {
+    fileprivate func box(_ value: Encodable) throws -> NSObject {
```
Slight nit: Do you mind keeping this method right above the implementation of `box_`? (i.e. move the dictionary version above this one)

itaiferber reviewed (changes requested):
This is getting really close! If you don't mind addressing some small nitpicking from me, I think we'll be good on this. :)

mortenbekditlevsen committed:
Fixes after comments from Itai

mortenbekditlevsen committed:
Added tests to ensure that coding key path are set correctly on encoding and decoding error contexts.

itaiferber reviewed (approved):
Looks great! Thanks for adding the new tests, too!

itaiferber commented:
@parkera Any thoughts on this?

itaiferber commented:
@swift-ci Please test macOS

itaiferber commented:
@swift-ci Please smoke test macOS

itaiferber commented:
@swift-ci Please smoke test Linux

parkera reviewed (approved):
Looks good in general.

Do we have a place to put release notes for this stuff?

itaiferber commented:
@swift-ci Please test macOS

itaiferber commented:
@parkera Unfortunately, not directly here, I don't think. We're going to need to mention this somewhere ourselves.

itaiferber commented:
The test failure doesn't look directly related here — looks like potential cause due to a different change. Retrying.

mortenbekditlevsen commented:
Should the tests be retried, or is there an underlying problem that needs fixing first?

harlanhaskins commented:
Might as well give this a run.

@swift-ci please test macOS platform

itaiferber commented:
It looks like the updated tests a revealing a real underlying issue. I haven't had time to investigate but I'll try to get this set up locally to figure out what's going on. It feels like something as high-level as this shouldn't be producing compiler failures.

mortenbekditlevsen commented:
Hi Itai,
Do you think there is any chance that this will be a part of Swift 4.2?

itaiferber commented:
@mortenbekditlevsen I'd really really like it to, but I've been swamped with other pressing work and haven't had time to investigate. I'd like to get to digging deeper here this week so we can make this failure actionable and try to get this in.

itaiferber commented:
@swift-ci Please test macOS

mortenbekditlevsen commented:
Hi @itaiferber ,
I'd like to give the bugfix a try. I'll start out from a clean master checkout and see if I can reproduce on the mac.
I'll let you know when I get stuck in the process. :-)

mortenbekditlevsen committed:
Merge branch 'master' into jsoncodingkeys

itaiferber commented:
@mortenbekditlevsen Excellent, thank you! For informative results here, you'll need to make sure you run the build script with the appropriate flags for building a debug version of the compiler.

If you don't already have this set up, one thing that might help get started is a config file to do this so you can avoid messing around with the build script flags; I've got the following:

```sh
$ cat ~/.swift-build-presets
[preset: minimal]
debug
assertions
stdlib-deployment-targets=macosx-x86_64
swift-stdlib-build-type=Debug
swift-stdlib-enable-assertions=1
skip-build-ios-device
skip-build-tvos-device
skip-build-watchos-device
skip-test-cmark
skip-test-sourcekit
skip-build-benchmarks

[preset: test]
mixin-preset=
    minimal

test
```

This allows me to run `./swift/utils/build-script --preset=minimal` and this builds just enough to get a debug compiler running on macOS, without spending the time to build for other platforms. Once you're set up there I can help you get the right invocation to run just a singular unit test file so you don't need to build all the tests.

mortenbekditlevsen commented:
@itaiferber Excellent - thanks! I had already run the build-script --test, so when running the minimal preset it completes really fast.
So how do I run the single failing unit test file?

itaiferber commented:
@mortenbekditlevsen Perfect. You can run a single test file with the `run-test` script; assuming you're in the root project directory (one level above the `swift` directory, and in the same place the `build` directory is), you should be able to run

```
./swift/utils/run-test --build-dir ./build/Ninja-DebugAssert ./swift/test/stdlib/TestJSONEncoder.swift
```

Once you've configured for running with tests, this should compile and run the single test file. Let me know if that works for you

mortenbekditlevsen commented:
@itaiferber it runs, but 'unfortunately' it does not fail. :-|
I cannot see the details of the execution above, so I don't know how it is 'supposed' to fail.

This ought to be reproducible on a standard mac, right?

mortenbekditlevsen commented:
@itaiferber I'm afraid that I can't get the test to fail. I verified that I am running with the latest master merged into the branch from this PR.
I wanted to have a look at what goes wrong on the CI server, but I can't seem to see the 'Details' links above (does the CI output expire?).

Do you have any suggestions on what could be different in my environment?
The error ought to be reproducible on a standard mac, right?

itaiferber commented:
@mortenbekditlevsen Looks like the CI output expired, unfortunately. Let me retrigger it so we can see what the failure should look like here. Did you compile the compiler in debug or release mode? If it's in debug mode, it's possible that this only reproduces in release mode...

itaiferber commented:
@swift-ci Please test macOS

mortenbekditlevsen commented:
@itaiferber yes, I did compile it in debug mode (figured that I would need to debug once the error is reproduced). I will try building in release mode!

mortenbekditlevsen commented:
@itaiferber Would you suggest building with --release og --release-debuginfo ?

mortenbekditlevsen commented:
@itaiferber I tried with --release-debuginfo and still the test completes without error. It will be interesting to see what's happening on the CI server. :-)

itaiferber commented:
Interesting...
@shahmishal, do you know off the top of your head what build settings we build with in CI? Seems strange to me that this issue wouldn't reproduce locally, but does in CI; I wonder what the differential there is.

itaiferber commented:
Okay, it looks like this failed in the 32-bit iOS simulator... That's likely why it doesn't reproduce on your Mac locally.

If you've got the appropriate Xcode with an iOS 12 simulator, you could build and test in the simulator environment, potentially.

mortenbekditlevsen commented:
@itaiferber Yay - I got it to fail:
`./swift/utils/run-test --build-dir ./build/Ninja-RelWithDebInfoAssert --target iphonesimulator-i386 ./swift/test/stdlib/TestJSONEncoder.swift`

Can I attach a debugger to the test executions? Until then I will make use of debug logging to test the difference between running with the simulator and without.

mortenbekditlevsen commented:
@itaiferber It appears that in the function `static bool doesNotRequireInstantiation(GenericWitnessTable *genericTable)` in `swift/stdlib/public/runtime/Metadata.cpp` the contents of `genericTable->Protocol->NumRequirements` is garbage when running in the simulator.
At least when I print the NumRequirements it appears to be containing values like 209059592.

I am trying to execute a single test as suggested in the output of running the test on a single file. But the built a.out binary is for the simulator, so the suggested command does not work. Do you know how I can execute the single binary in the simulator environment?
```/Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/test-iphonesimulator-i386/stdlib/Output/TestJSONEncoder.swift.tmp/a.out --stdlib-unittest-in-process --stdlib-unittest-filter "testEncodingTopLevelStructuredSingleClass"
dyld: mach-o, but built for simulator (not macOS)
Abort trap: 6
```

mortenbekditlevsen commented:
@itaiferber Ok, I have a few more findings. I am debugging using `fprintf` and rebuilding so it's going a bit slow. Also I don't really know how to set everything up in Xcode, so every time I encounter a new type I spend quite a bit of time grepping through the code base to figure out the type details.

But it appears that in the `doesNotRequireInstantiation` function, the genericTable->Protocol contains garbage when I run it in the simulator. At least the `NumRequirements` looks like garbage and I get a crash when I try to print the name. 

Running the same tests without the simulator, I can see that the protocol in question is the `_JSONStringDictionaryDecodableMarker` as expected.

I did try making the protocol internal to check if the error had to do with it being `fileprivate`, but it still fails when it's internal.
 
I am a bit uncertain about how to proceed. I have tried to figure out where the GenericWitnessTable containing the garbage protocol gets created, but I haven't managed to figure it out. Since we are dealing with the runtime I don't know how much of the data comes from pre-generated code and how much is actually created at runtime.

Do you have any suggestions for continuing the debugging? :-)

mortenbekditlevsen commented:
@itaiferber I managed to connect a debugger to the simulator, so now I am comparing stack traces between sim and native. Big improvement to have the debugger, but still a bit uphill. ;-)

itaiferber commented:
Thanks for taking a deep dive into this! Witness table generation issues are a bit out of my scope, but we can help dig deeper. @jrose-apple, who would best be able to help debug witness table errors in the 32-bit iOS simulator? (For a quick reference, this PR adds a `fileprivate` marker protocol to identify `Dictionary<String, ...>`, and attempting to use this protocol internally ends up crashing in the 32-bit iOS simulator with what appears to be a garbage witness table.)

jrose-apple commented:
@jckarter, maybe?

mortenbekditlevsen commented:
@jckarter @itaiferber I have tried comparing stacktraces between the simulator environment and the non-simulator environment for the frame where the assertion happens, but the traces seem to differ quite a bit, so I am a bit confused about that - and a bit uncertain about how to proceed with the debugging.

The error appears to be triggered in line 882 of the `JSONEncoder.swif` file when an `is` check is made in the `box_` function:

```
        } else if value is _JSONStringDictionaryEncodableMarker {
```

Any suggestions would be greatly appreciated - specifically about where the `GenericWitnessTable` structure is being initialized.

itaiferber commented:
Unfortunately, I know nothing about witness table generation, but maybe with some more information I can help point the way. If @jckarter has time he might be able to weigh in too.

Can you share more information about the failure? Error messages/output would be best.

jckarter commented:
@slavapestov would know more about how those structures work now than I do. Instead of focusing on the data structure itself, though, I would look at the code being generated at the LLVM IR level. If we're seeing garbage values passed down, there's possibly a bug in the way we form the calling convention for a generic environment somewhere.

mortenbekditlevsen commented:
@jckarter Ok, I would like to try looking in to that. Do I need to enable SIL generation for the stdlib or am I confusing SIL and IR here?
I am running the following command to provoke the error:
```
./swift/utils/run-test --build-dir ./build/Ninja-RelWithDebInfoAssert --target iphonesimulator-i386  ./swift/test/stdlib/TestJSONEncoder.swift
```
I assume that the error comes from the change to `stdlib/public/SDK/Foundation/JSONEncoder.swift`so that is why I am guessing that I need to enable SIL generation for the stdlib.

jckarter commented:
SIL is probably still too abstract to see the exact problem here. I would look at the LLVM IR.

mortenbekditlevsen commented:
@jckarter Unfortunately I haven't been able to figure out how to output LLVM IR while building the stdlib. Any pointers would be greatly appreciated!

jckarter commented:
I don't think there's an automated way to do it. You can `ninja -v`, take the commandline ninja used to build the Swift standard library, and remove any `-c` or `-o Swift.o` options out of the commandline and replace them with `-emit-ir` (or `-emit-sil` if you want to see SIL output)

mortenbekditlevsen commented:
@itaiferber @jckarter Ok, now I succeeded in generating the IR for Foundation on x86_64. I removed the old `Foundation.o` file and ran `ninja -v` as you suggested, @jckarter .

mortenbekditlevsen commented:
The command that `ninja -v` showed me is:

```
cd /Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/stdlib/public/SDK/Foundation && /usr/bin/python /Users/mortenbekditlevsen/OpenSource/2/swift/utils/line-directive @/Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/stdlib/public/SDK/Foundation/mwhso.txt -- /Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/./bin/swiftc -c -sdk /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk -target x86_64-apple-macosx10.9 -resource-dir /Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/./lib/swift -F /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/../../../Developer/Library/Frameworks -O -g -D INTERNAL_CHECKS_ENABLED -D SWIFT_ENABLE_RUNTIME_FUNCTION_COUNTERS -I /Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/./lib/swift/macosx/x86_64 -module-cache-path /Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/./module-cache -no-link-objc-runtime -Xfrontend -enable-resilience -swift-version 3 -swift-version 4 -autolink-force-load -warn-swift3-objc-inference-complete -Xfrontend -verify-syntax-tree -Xllvm -sil-inline-generics -Xllvm -sil-partial-specialization -swift-version 3 -Fsystem /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/PrivateFrameworks/ -module-link-name swiftFoundation -force-single-frontend-invocation -parse-as-library -o /Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/stdlib/public/SDK/Foundation/macosx/x86_64/Foundation.o @/Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/stdlib/public/SDK/Foundation/mwhso.txt
```

My modification to generate x86_64 IR output is:

```
cd /Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/stdlib/public/SDK/Foundation && /usr/bin/python /Users/mortenbekditlevsen/OpenSource/2/swift/utils/line-directive @/Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/stdlib/public/SDK/Foundation/mwhso.txt -- /Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/./bin/swiftc -emit-ir -sdk /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk -target x86_64-apple-macosx10.9 -resource-dir /Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/./lib/swift -F /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/../../../Developer/Library/Frameworks -O -g -D INTERNAL_CHECKS_ENABLED -D SWIFT_ENABLE_RUNTIME_FUNCTION_COUNTERS -I /Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/./lib/swift/macosx/x86_64 -module-cache-path /Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/./module-cache -no-link-objc-runtime -Xfrontend -enable-resilience -swift-version 3 -swift-version 4 -autolink-force-load -warn-swift3-objc-inference-complete -Xfrontend -verify-syntax-tree -Xllvm -sil-inline-generics -Xllvm -sil-partial-specialization -swift-version 3 -Fsystem /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/PrivateFrameworks/ -module-link-name swiftFoundation -force-single-frontend-invocation -parse-as-library -o /Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/stdlib/public/SDK/Foundation/macosx/x86_64/Foundation.ir @/Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/stdlib/public/SDK/Foundation/mwhso.txt
```

My attempt at running this through `xcrun` is:
```
cd /Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/stdlib/public/SDK/Foundation && /usr/bin/python /Users/mortenbekditlevsen/OpenSource/2/swift/utils/line-directive @/Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/stdlib/public/SDK/Foundation/mwhso.txt -- xcrun --toolchain default --sdk /Applications/Xcode-beta.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator12.0.sdk /Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/./bin/swiftc -emit-ir -sdk /Applications/Xcode-beta.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk -target i386-apple-ios7.0 -resource-dir /Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/./lib/swift -F /Applications/Xcode-beta.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/../../../Developer/Library/Frameworks -O -g -D INTERNAL_CHECKS_ENABLED -D SWIFT_ENABLE_RUNTIME_FUNCTION_COUNTERS -I /Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/./lib/swift/iphonesimulator/i386 -module-cache-path /Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/./module-cache -no-link-objc-runtime -Xfrontend -enable-resilience -swift-version 3 -swift-version 4 -autolink-force-load -warn-swift3-objc-inference-complete -Xfrontend -verify-syntax-tree -Xllvm -sil-inline-generics -Xllvm -sil-partial-specialization -swift-version 3 -Fsystem /Applications/Xcode-beta.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/ -module-link-name swiftFoundation -force-single-frontend-invocation -parse-as-library -o /Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/stdlib/public/SDK/Foundation/iphonesimulator/i386/Foundation.ir @/Users/mortenbekditlevsen/OpenSource/2/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/stdlib/public/SDK/Foundation/mwhso.txt
```

which is the command that gives me the error:
`<unknown>:0: error: unable to load standard library for target 'i386-apple-ios7.0'`

So any hints at how to run the `ninja` command for the i386-iphonesimulator target would be awesome. :-)

mortenbekditlevsen commented:
@itaiferber @jckarter I have now been able to generate IR for the Foundation module for i386/iphonesimulator as well as for x86_64. I am comparing the two and looking for anything related to witness tables for the marker protocols. But I don't know what I am searching for, so I am a bit stuck.

Would it help if I share some of the IR code?
Sincerely,
/morten

jckarter commented:
Sure, you could paste them here or in gists.

mortenbekditlevsen commented:
@jckarter 
Here is a gist containing the references to the _JSONStringKeyedDictionaryEncodableMarker protocol on i386 / simulator and x86_64 respectively.
I have left out a few references - let me know if you would like everything.

https://gist.github.com/mortenbekditlevsen/563013b3cb12b6093fcdb91a2f3a5691

jckarter commented:
Sorry, I haven't had a chance to look at this. Another IRGen expert like @aschwaighofer or @rjmccall might be able to help too. In the meantime, filing a PR would also help draw more attention to the problem. Sorry again!

mortenbekditlevsen commented:
There’s really nothing to apologize for :-) - I am only pushing the issue a bit since I would love for this PR to become part of the final 4.2.
By ‘filing a PR’ do you mean a Swift bug report?

jckarter commented:
Sorry, I meant bug report, yeah.

mortenbekditlevsen commented:
Super - I've added a bug report with a reference to this PR:
https://bugs.swift.org/browse/SR-8276

itaiferber commented:
One more tentative test, but I don’t expect much to have changed.

itaiferber commented:
@swift-ci Please test

itaiferber commented:
@jckarter @jrose-apple @slavapestov This is something we’d really like to get into 4.2 if we can (the sooner we fix this the less breaking the change will be) — is there anything creative you folks can think of that we can do to avoid hitting this, or a way for us to push forward here?

It’s understandable if this has to wait until Swift 5 for us to investigate SR-8276, but if we can get at the generic type some other way, that’d be great to explore.

mortenbekditlevsen committed:
Merge branch 'master' into jsoncodingkeys

mortenbekditlevsen committed:
Merge remote-tracking branch 'upstream/master' into jsoncodingkeys

mortenbekditlevsen commented:
@itaiferber, @jckarter - in case the issue with dynamic casting of conditional conformances is not fixed before 4.2 is released, should I try coming up with a fix for the original problem that does not require the marker protocol? It can be done by adding some state I assume.

mortenbekditlevsen commented:
Hi again @itaiferber , @jckarter 
I may have an acceptable workaround. If the marker protocols are made 'internal' instead of 'fileprivate', it appears that the bug is not being triggered.
Would this be an acceptable workaround - if supplied with a comment about why the access modifier is changed?

jckarter commented:
That seems like a reasonable workaround.

mortenbekditlevsen committed:
Suggestion for a workaround for runtime error related to witness tables on i386

mortenbekditlevsen commented:
I have committed the small change. Looking forward to seeing if the CI is satisfied. :-)
Is there a way for me to annotate the change with a reference to the bug I reported, so that the change can be reverted once the bug is fixed?

itaiferber commented:
@swift-ci Please test

itaiferber commented on stdlib/public/SDK/Foundation/JSONEncoder.swift:
```diff
@@ -10,6 +10,35 @@
+///
+/// The marker protocol also provides access to the type of the `Decodable` values,
+/// which is needed for the implementation of the key conversion strategy exemption.
+#if arch(i386) || arch(arm)
```
I'd leave a comment here with a link to the SR you filed regarding this. Also, since this only affected the 32-bit iOS simulator, I think we can cut down on this `#if` a bit in practice, right?

itaiferber replied:
In fact, this might be a good candidate for `#if targetEnvironment(simulator)`

mortenbekditlevsen committed:
Added comment and included reference to bug

mortenbekditlevsen commented:
@itaiferber thanks a lot - I've added a comment with a reference to the SR and changed the checks to specifically target i386 simulator.

I looked at the failing CI builds, but I can't seem to recognize the reported issues.
Can you determine if the CI fails because of some outside issue or if I have introduced something bad?

itaiferber commented:
@mortenbekditlevsen I don't remember under exactly what conditions CI hits the "invalid SHA" but it always re-checks out the branch and re-runs the tests after the "failure". I think it likely re-ran but since you pushed other changes since, those results have been lost. I'll re-run now so we can see where we are.

itaiferber commented:
@swift-ci Please test macOS

itaiferber commented:
@swift-ci Please smoke test Linux

itaiferber commented:
@mortenbekditlevsen Unfortunate but real failure here:

```
/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/swift/stdlib/public/SDK/Foundation/JSONEncoder.swift:886:20: error: variable used within its own initial value
11:47:53         let type = type(of: value)
```

This is a separate issue that I think should be remedied (likely worth a JIRA — I'll look it up), but Swift thinks that the LHS `type` is the same as the RHS `type`. This can be remedied either by changing `type(of: value)` to `Swift.type(of: value)` to disambiguate, or by renaming `let type` to `let valueType` or similar.

itaiferber commented:
For the above: https://bugs.swift.org/browse/SR-1846

mortenbekditlevsen committed:
Disambiguation between variable and function with same name

mortenbekditlevsen commented:
@itaiferber Thanks - I think it must be a regression because it runs on my locally checked out version (which is a few weeks old). I am adding the disambiguation with a comment referring SR-1846. Just running the tests locally before I push the change. :-)

Should be up in 15 minutes!

mortenbekditlevsen commented:
@itaiferber The disambiguation is now pushed to the PR.

itaiferber commented:
@swift-ci Please test macOS

itaiferber commented:
@swift-ci Please smoke test Linux

itaiferber commented:
@mortenbekditlevsen Thanks!

mortenbekditlevsen commented:
Yay - the lights! They are all green!!! ❤️So happy! 😊

So this PR is against master. Will master be merged to 4.2 again or should I make a new PR for the 4.2 branch? Or should the merge commit be cherrypicked to 4.2?
@itaiferber

itaiferber commented:
@mortenbekditlevsen We would need to cherry-pick this to the branch that we want, but unfortunately, [the only changes we can take for swift-4.2 specifically right now are critical ones](https://forums.swift.org/t/swift-4-2-in-final-convergence-swift-4-2-branch-open-for-post-4-2-0-changes/15128). Changes being made to `swift-4.2-branch` right now will go into a potential future `swift-4.2.1` release or similar. @tkremenek Are there guidelines right now for what is appropriate for `swift-4.2.1` vs. a future release?

itaiferber commented:
@mortenbekditlevsen And BTW, thanks for taking this on and sorry this has dragged on for so long!

mortenbekditlevsen commented:
@itaiferber You are very welcome - I'm happy to be able to contribute! :-)
Of course a bit sad that the bug fix will not make it into 4.2, but I understand the reasons.
Thank you very much for your guidance!
